torch/csrc/distributed/rpc/utils.cpp

```diff
@@ -272,12 +271,13 @@ std::string wireSerialize(
     pickler.pushIValue(cloneSparseTensors(tensors));
     pickler.stop();
     // tensorData is in function scope so that the data() pointers stay valid.
```

Comment:
update this comment too?

Reply:
I think what this comment tried to say was that we need to keep the WritableTensorData objects alive until lines 298-300, as that is the moment where we actually copy the data from the tensor to the buffer that will be sent over the network. The reason for this, I think, is because there could be temporary tensors that are only kept alive from WritableTensorData and would be deallocated if WritableTensorData is destroyed. I see two such types of tensors:
- the ones created by cloneSparseTensors in case we're only sending a small slice of a larger tensor;
- the ones created by getWritableTensorData in case the original tensor was a CUDA tensor.

Since the ProcessGroup agent doesn't support CUDA tensors I think the only real case where this could happen is the first one. And in that case I think this diff would introduce a bug. I'm a bit surprised there is no test that caught that.

Reply:
The `Ent` instances in the `entries` vector just contain raw pointers to the underlying data of the tensor. That data is however owned and managed by an `at::Tensor` instance and as soon as it goes out of scope it deallocates the memory, meaning that the raw pointer in the `Ent` becomes invalid and may lead to a page fault when used. We thus need to keep the `at::Tensor` instances alive as well until we have copied their data out of them. For most tensors that will be automatic because they are kept alive by the caller of this function. However, there are also temporary `at::Tensor` instances created within this function, namely the ones created by `cloneSparseTensors`. Those need to be stored somewhere in order to keep them alive. This is what the comment is saying. If you feel the wording could be improved go ahead! :)

Reply:
gotcha! thanks for the detailed explaination, I tried locally, this indeed trigger a bug with the `cloneSparseTensors` case, and it's captured by https://github.com/pytorch/pytorch/blob/fe88806784e05af554247c8a29467adb6d84cc5b/test/cpp/rpc/test_wire_serialization.cpp#L39

But apparently this cpp test is not run in CI by default, will need to figure out how to enable this on CI.